### ISSUE TITLE
KITTI Results

### ISSUE THREAD

**Issue by CuongNguyen218** [AlexeyAB/darknet#4454]
@AlexeyAB ,
Hi Alex. Thanks for your amazing code. But I have a trouble when Training your repo with KITTI Dataset.
When I use your repo and trained around 50k iter, I get a bad result with car AP detection when test with KITTI Devkit. Details in image below.
![image](https://user-images.githubusercontent.com/43350150/70254231-53c38d00-17b7-11ea-8588-46a2b50d5692.png)

And my result at final iter with your repo below
![image](https://user-images.githubusercontent.com/43350150/70246201-eceba700-17a9-11ea-89d0-ffcfc81b3004.png)
And chart image of training process below:
![image](https://user-images.githubusercontent.com/43350150/70246282-160c3780-17aa-11ea-90d6-aa447219dd81.png)
I see a gap between the result of car mAP when I use your repo and GaussianYolo repo is large. Notice that I don't use Gaussian Yolo Layer. Result from Gaussian Yolo repo in the image below
![image](https://user-images.githubusercontent.com/43350150/70246547-83b86380-17aa-11ea-899c-403d2ec964f8.png)
Can you explain it for me ? 
This is my config file
[yolov3-kitti.txt](https://github.com/AlexeyAB/darknet/files/3927532/yolov3-kitti.txt)
:
And this is the number of box of KITTI dataset in training set and test set:
![image](https://user-images.githubusercontent.com/43350150/70247057-4b655500-17ab-11ea-972b-d5743c15be14.png)

**Comment by AlexeyAB** [comment]
@CuongNguyen218 Hi,

1. There was a bug in the data augmentation from 26 oct to 4 dec 2019. So you should train it again in this repo.

2. Also what Darknet-repo and what command do you use to get car_detection AP?

**Comment by CuongNguyen218** [comment]
@AlexeyAB , Hi,
Thank for your quick reply. Darknet repo what i use to reproduce this result below is Gaussian Yolo Repo ( i used yolo layer not gaussian Yolo Layer).
![image](https://user-images.githubusercontent.com/43350150/70251195-a4d08280-17b1-11ea-9270-67d31adfa27c.png)
Gaussian Repo: https://github.com/jwchoi384/Gaussian_YOLOv3
and I used official evaluate code from KITTI dataset webpage without any change: http://www.cvlibs.net/datasets/kitti/eval_object.php?obj_benchmark=2d

**Comment by AlexeyAB** [comment]
> Darknet repo what i use to reproduce this result below is Gaussian Yolo Repo

Try to use this repo: https://github.com/AlexeyAB/darknet

I just added `./darknet detector valid` for KITTI and BDD: https://github.com/AlexeyAB/darknet/commit/aaba21bdd336c77561f6c79cb940ef9b52a2b11b

**Comment by CuongNguyen218** [comment]
@AlexeyAB ,
thank for your support but I think there is a misunderstanding between us. Let me explain:

1. I trained KITTI dataset with your repo get weight file A. Then, I  use weight file A to valid on testset and result of weight A in gaussian repo and get results in image below when test with devkit code
![image](https://user-images.githubusercontent.com/43350150/70253580-0c88cc80-17b6-11ea-8f85-605ad1f755f6.png)
I see it is an unreasonable result follow gaussian paper: 
![image](https://user-images.githubusercontent.com/43350150/70253690-3d690180-17b6-11ea-916e-0df1d3134f5b.png)

2. To check problems, I retrained KITTI dataset with Gaussian Yolov3 Repo and get weight file B (it don't have GIOU value in log file and I don't know that GIOU affects to this result when i use your  repo). Then, I used weight B to valid on testset and get results in image below with devkit code.
![image](https://user-images.githubusercontent.com/43350150/70253991-e879bb00-17b6-11ea-8d1c-05139b138442.png)

Maybe i will retry your newest repo. Thanks

**Comment by AlexeyAB** [comment]
> My cmd line is: ./darknet detector valid cfg/datafile cfg/cfgfile weights
> Maybe i will retry with your newest repo. Thanks

1. Yes.  You should do `./darknet detector valid` on the same repo where did you do `./darknet detector train` There are different resize approaches in different frameworks with pros and cons: https://github.com/AlexeyAB/darknet/issues/232#issuecomment-336955485

2. Also there was a bug 26 oct -> 4 dec in training: https://github.com/AlexeyAB/darknet/issues/4454#issuecomment-562188010

**Comment by CuongNguyen218** [comment]
@AlexeyAB , 
I will retrain tomorrow. Thanks

**Comment by CuongNguyen218** [comment]
@AlexeyAB , 
Hi i want to read about data augment bug. Maybe you prefer wrong comment above?

**Comment by AlexeyAB** [comment]
> I will retrain tomorrow. Thanks you. One more question, I want to use GIOU with Gaussian Yolo Layer as a metrics instead of IOU ( do not use GIOU loss). How can I do that ?

There isn't mAP that is based on GIoU instead of IoU. Try to ask it in GIoU-repository.

> Hi i want to read about data augment bug. Maybe you prefer wrong comment above?

Just Bilateral-blurring will be used for 50% of images even if you didn't enable it.

**Comment by CuongNguyen218** [comment]
@AlexeyAB ,
In #3114 in###  yolo_v3_tiny_pan5 matrix_gaussian_GIoU aa_ae_mixup_new.cfg , I saw you use GIOU for Gaussian Yolo layer. I think something not correct when see NLL loss in Gaussian Yolo Paper because NLL loss in Gaussian Yolo Layer is loss function for bounding box and GIOU also Bounding Box loss. Can you explain it for me ?

**Comment by AlexeyAB** [comment]
@CuongNguyen218 

GIoU just replaces NLL loss for b-box.
So there will be GIoU-loss for bbox-loss and Gaussian_yolo-loss for Uncertainty.

**Comment by CuongNguyen218** [comment]
@AlexeyAB ,
Hi. I think something wrong with valid
![image](https://user-images.githubusercontent.com/43350150/70373571-9c945680-191b-11ea-8266-bb146173c889.png)
I retrained KITTI dataset with your new code and your valid code but something when wrong with car class. The trend is Easy > Moderate > Hard but my result what I get from your code is not follow that.
I don't know why ?

**Comment by AlexeyAB** [comment]
@CuongNguyen218 

Also you can try to set in your cfg-file
```
[net]
letter_box=1
```

Then train usin AlexeyAB repo, and valid by using Jwchoi repo.

**Comment by CuongNguyen218** [comment]
@AlexeyAB , Thanks for your support. and i want to find commit of bag of freebie. Can you find it for me?

**Comment by AlexeyAB** [comment]
Serach commit with MixUp.

or just use
mixup=1

**Comment by CuongNguyen218** [comment]
@AlexeyAB , i saw `bag_freebies=1` in #3272 . does it equal to `mixup = 1` ?

**Comment by AlexeyAB** [comment]
Yes.

**Comment by CuongNguyen218** [comment]
Thanks for

**Comment by AlexeyAB** [comment]
But you should know, that mixup=1 improves accuracy only for Classifier. (and the use Classifier pre-trained weights file for Detector).


For training Detector is recommended to use `mosaic=1`

**Comment by CuongNguyen218** [comment]
okay, I got it. Thanks you. I want to try it because i see it improve performance of yolov3 in gluoncv +5% on COCO.
Paper: https://arxiv.org/pdf/1902.04103.pdf
Maybe it include a bag of trick for object detection.

**Comment by AlexeyAB** [comment]
BoF (Bag of Freebies) includes 5 features - there are implemented 4 of 5 features in this repository, read: https://github.com/AlexeyAB/darknet/issues/3272#issuecomment-497149618

There is implemented `mixup=1` that gives `+1.5 - +2.0% mAP` for Yolov3, but:

`mixup=1` for Detector gives **only +0.3%** mAP@0.5...0.95 on MS COCO

Paper: https://arxiv.org/pdf/1902.04103.pdf
![image](https://user-images.githubusercontent.com/4096485/70379977-75eb1580-1945-11ea-9131-1421a1c053f2.png)

----

To get `+1.5 - +2.0% mAP` you must first train the Classifier with `[net] mixup=1` https://github.com/AlexeyAB/darknet/blob/master/cfg/darknet53.cfg  and then use this pre-trained weights file for training Detector Yolo with `[net] mixup=1` https://github.com/AlexeyAB/darknet/blob/master/cfg/yolov3.cfg then you will get **+2% MS COCO**

----

To get another +2.0% you should use: mixup=1, SGDR (cosine lr), random=1, and GPU-processing + CPU-RAM - it takes a lot of time to training.

----

There are better approaches than MixUP:
* CutMix: https://github.com/AlexeyAB/darknet/issues/4419 - for Classifier
* Mosaic: https://github.com/AlexeyAB/darknet/issues/4264 - for Classifier and Detector

**Comment by CuongNguyen218** [comment]
@AlexeyAB , 
I retrained my model with your repo and setting `letter_box = 1` in my cfg file, and bad results for car class still appear.
This is my result:
![image](https://user-images.githubusercontent.com/43350150/70390230-b4430c00-19fb-11ea-9461-906760c850ab.png)

**Comment by AlexeyAB** [comment]
What width= and height= do you use in both cfg-files?
What random= do you use in the last [yolo] layer in both cfg-files?

**Comment by CuongNguyen218** [comment]
Hi @AlexeyAB ,
I set width = height = 512 in both cfg file.
random = 1, ignore_thresh = 0.5 in both.

**Comment by AlexeyAB** [comment]
> I retrained my model with your repo and setting letter_box = 1 in my cfg file, and bad results for car class still appear.

Did you set `[net] letter_box=1` in cfg, then train with AlexeyAB, then valid with Jwchoi and get Car  AP = 59.56%?

Can you attach both cfg-files?

There is also different network resize approaches:
* AlexeyAB: https://github.com/AlexeyAB/darknet/blob/318919e1cbb362aac6cb5c3d9388735f9ab594b6/src/detector.c#L169-L171
* pjreddie that is doesn't depend on the initial network size: https://github.com/pjreddie/darknet/blob/61c9d02ec461e30d55762ec7669d6a1d3c356fb2/examples/detector.c#L65

**Comment by CuongNguyen218** [comment]
Did you set [net] letter_box=1 in cfg, then train with AlexeyAB, then valid with Jwchoi and get Car AP = 59.56%?
Yes.
Here is my both cfg-files?
[yolov3-kiit-jwchoi.txt](https://github.com/AlexeyAB/darknet/files/3936345/yolov3-kiit-jwchoi.txt)
[yolov3-kitti-AlexeyAB.txt](https://github.com/AlexeyAB/darknet/files/3936348/yolov3-kitti-AlexeyAB.txt)

**Comment by AlexeyAB** [comment]
Did you train both models with the same batch and subdivisions?
```
batch=64
subdivisions=16
```

Try to change these lines: https://github.com/AlexeyAB/darknet/blob/318919e1cbb362aac6cb5c3d9388735f9ab594b6/src/detector.c#L169-L171

To these
```cpp
            int dim = (rand() % 10 + 10) * 32;
            if (get_current_batch(net)+200 > net->max_batches) dim = 608;
int dim_w = dim;
int dim_h = dim;
```


recompile and train **without** letter_box=1

**Comment by CuongNguyen218** [comment]
> Did you train both models with the same batch and subdivisions?
> 
> ```
> batch=64
> subdivisions=16
> ```
Yes, I set both are Batch=64 and sub =16
i will retry this with 10000k iter.

**Comment by CuongNguyen218** [comment]
@AlexeyAB ,
I change these line above and get an error

**Comment by AlexeyAB** [comment]
Also remove these lines: https://github.com/AlexeyAB/darknet/blob/318919e1cbb362aac6cb5c3d9388735f9ab594b6/src/detector.c#L173-L177

**Comment by AlexeyAB** [comment]
Use `if (get_current_batch(net) + 200 > net.max_batches) dim = 608;` instead of `if (get_current_batch(net) + 200 > net->max_batches) dim = 608;`


Also removes these lines: 
https://github.com/AlexeyAB/darknet/blob/318919e1cbb362aac6cb5c3d9388735f9ab594b6/src/detector.c#L173-L177

**Comment by CuongNguyen218** [comment]
![image](https://user-images.githubusercontent.com/43350150/70391384-2705b480-1a07-11ea-944e-ea661f4a60cd.png)
This is new error.
![image](https://user-images.githubusercontent.com/43350150/70391393-3ab11b00-1a07-11ea-9401-9800bf1f3390.png)
Here is my change in cođe. Maybe I miss a '(' right?

**Comment by AlexeyAB** [comment]
Use `//` for commenting instad of `#`

**Comment by CuongNguyen218** [comment]
@AlexeyAB , thank you it work. As I feel, training speed is faster.

**Comment by CuongNguyen218** [comment]
@AlexeyAB ,
I trained my model with 10k iter and finally it works.
![image](https://user-images.githubusercontent.com/43350150/70418599-4e579280-1a96-11ea-80c2-1246dc0bcf6c.png)
Thank you for your support.

**Comment by err-or** [comment]
Hi, if you are using `./darknet detector valid ` for kitti, what do you use for `eval=` in obj,data? If you are using the devkit to evaluate yolo for kitti, can you please tell me how you are using the devkit? Thanks!

**Comment by AlexeyAB** [comment]
Use `eval=kitti` in the `obj.data` file.

Can be: voc, coco, imagenet, kitty, bdd: https://github.com/AlexeyAB/darknet/blob/49bff0e24da36f7c1a2471b12089efb27a0173b1/src/detector.c#L662-L708

**Comment by err-or** [comment]
Hi Alexey, thank you for your response. I did what you said but I get an error while running `./darknet detector valid` where `eval = kitti` in `obj.data`. The error is this: 
![image](https://user-images.githubusercontent.com/61436133/89117270-83bc5080-d46a-11ea-8636-d54eac0e032a.png)
To give you an idea of what I did: 
1. I downloaded the training images of KITTI 2D object detection
2. I have converted the labels according to yolo
3. I have taken around 6k images as train.txt
4. The remaining images are in valid.txt
5. I trained the model up to 16k iterations

Can you please tell me how can I evaluate the model according to kitti 2D detection benchmark?

**Comment by AlexeyAB** [comment]
* Can you detect successfully by using your model?

* do you have `/results/` folder?

* do you have folder `/results/kitti_results/data/` ?

* show `obj.data` file, did you set `eval=kitti` ?

* show command that you used

* Can you successfully run `./darknet detector map ...` on kitti val dataset?


----

If you can't get Kitti-evalutation, also you can try to use BDD-evalutation methond

* eval=bdd in obj.data file

*  `./darknet detector valid cfg/kitty.data cfg/kitty.cfg kitty.weights`

* try to download `gt_KITTI_valid_for_bdd_eval_method.json` from https://github.com/jwchoi384/Gaussian_YOLOv3/tree/master/bdd_evaluation

* and try to use this script https://github.com/jwchoi384/Gaussian_YOLOv3/blob/master/bdd_evaluation/evaluate.py
```
cd bdd_evaluation/

python evaluate.py det gt_KITTI_valid_for_bdd_eval_method.json ../results/bdd_results.json
```

**Comment by err-or** [comment]
Hi Alexey, 
1. Yes I can successfully detect with my model
2. The `/results/` folder was already there even before running the training
3. I get the `/results/kitti_results/data/` folder when I run `./darknet detector valid` with `eval = kitti` in `obj.data` but I encounter the error that I mentioned. I only get a single .txt file one of the image file but that .txt file is blank.
4. This is the command that I used: `./darknet detector valid data/obj.data cfg/yolo-obj.cfg backup/yolo-obj_final.weights` 
5. Yes I can successfully run `./darknet detector map ...` and this is what I get after running this: 

![image](https://user-images.githubusercontent.com/61436133/89135167-83b76180-d4f9-11ea-9095-28966fc996c7.png)
